### PR TITLE
Implement letterSpacing on Android >= 5.0

### PR DESCRIPTION
**Pull request by motiz88**
## Motivation

`letterSpacing` is completely missing from RN Android at the moment.

I've reviewed the `letterSpacing` implementations in #13199, #13877 and #16801 (that all seem to have stalled) and managed to put together an improved one based on #13199, updated to merge cleanly post https://github.com/facebook/react-native/commit/6114f863c3aed826355530bcf404ee5aed2f927b, that resolves the [issues](https://github.com/facebook/react-native/pull/13199#issuecomment-354568863) I've identified with that code.

I believe this is the closest PR yet to a correct implementation of this feature, with a few caveats:

- As with the other PRs, this only works on Android >= 5.0 (silently falling back to no letter spacing on older versions). Is this acceptable for a RN feature, in general? Would a dev mode warning be desirable?
- The other PRs seem to have explored the space of potential solutions to the layout issue ([Android renders space _around_ glyphs](https://issuetracker.google.com/issues/37079859), iOS to the _right_ of each one) and come up empty, so I've opted to merely document the difference.
- I have neither updated nor tested the "Flat" UI implementation - everything compiles but I've taken [this comment](https://github.com/facebook/react-native/issues/12770#issuecomment-294052694) to mean there's no point in trying to wade through it on my own right now; I'm happy to tackle it if given some pointers.
- The implementation in `ReactEditText` is only there to handle the placeholder text, as `ReactBaseTextShadowNode` already affects the input control's contents correctly.
  - I'm not sure whether `<TextInput>` is meant to respect `allowFontScaling`; I've taken my cue here from `ReactTextInputManager.setFontSize()`, and used the same units (SP) to interpret the value in `ReactEditText.setLetterSpacingPt()`.
  - I'm not sure whether `<TextInput>` is even meant to support `letterSpacing` - it doesn't actually work on iOS. I'm not going to be able to handle the Objective-C side of this, not as part of this PR at least.
- I have not added unit tests to `ReactTextTest` - is this desirable? I see that some other props such as `lineHeight` aren't covered there (unless I'm not looking in the right place).
- Overall, I'm new to this codebase, so it's likely I've missed something not mentioned here.

## Test Plan

Note comment re: unit tests above; RNTester screenshots follow.

### RNTester - `<Text>`
| iOS (existing functionality, amended test) | Android (new functionality & test) |
| - | - |
|  |  |

### RNTester - `<TextInput placeholder>`
| iOS _(not implemented, test not in this branch)_ | Android (new functionality & test) |
| - | - |
|  |  |

### RNTester - `<TextInput value>`
| iOS _(not implemented, test not in this branch)_ | Android (new functionality & test) |
| - | - |
|  |  |

## Related PRs

https://github.com/facebook/react-native-website/pull/105 - this docs PR is edited slightly from what's in `TextStylePropTypes` here; happy to align either one to the other after a review.

## Release Notes

[ANDROID] [FEATURE] [Text] - Implemented letterSpacing

### CODE REVIEW

**Comment by janicduplessis** [comment]
Thanks for the PR, going to try it in my RN fork to see if it works properly in some apps.

@hramos Could you have someone review this, not sure who to ping for android these days :S

**Comment by janicduplessis** [comment]
@motiz88 Tested this and there is one pretty major issue. Text isn't measured properly when using letterSpacing. It can be made very noticeable by adding a background color.

Letter spacing 2.0:
image

No letter spacing:
image

**Comment by motiz88** [comment]
@janicduplessis Ooh, that's going to be interesting to track down. On it, will report back soon.

**Comment by motiz88** [comment]
Can you share some code to reproduce the issue you're seeing, @janicduplessis? Trying this in RNTester (see code in 58611cd9a105a3195de6eeb35c6004f635725604) seems to render correctly. (The bottom row isn't wrapped with `flexDirection: row` so gets full width - as I understand, this is per the standard RN layout algorithm for text)

| iOS | Android |
|-|-|
|  |   |

**Comment by janicduplessis** [comment]
@motiz88 Oh, looks like it only happens when setting fontSize + letterSpacing.

This repros it for me:
```js
<View style={{ flex: 1, alignItems: 'flex-start' }}>
        <Text
          style={{
            backgroundColor: 'red',
            letterSpacing: 2,
            fontSize: 12,
          }}
        >
          TEST TEST TEST
        </Text>
      </View>
```

**Comment by motiz88** [comment]
Fixed in 6408a25. Apparently, spans end up being evaluated in reverse order to the one in which they're `add`ed to the `ops` list in `buildSpannedFromShadowNode` - probably due to [this line](https://github.com/motiz88/react-native/blob/6408a25d101a5f514cf17cb32c78df564e692f07/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactBaseTextShadowNode.java#L216) increasing each subsequent one's priority, if I'm reading the code correctly. This in turn [matters to `CustomLetterSpacingSpan`](https://github.com/motiz88/react-native/blob/6408a25d101a5f514cf17cb32c78df564e692f07/ReactAndroid/src/main/java/com/facebook/react/views/text/CustomLetterSpacingSpan.java#L23-L25), as it reads the current font size from the `TextPaint` instance to perform an appropriate unit conversion.

Quite an omission in my testing approach 😨  especially as I was very much aware of the dependency between `fontSize` and `letterSpacing`. But looks good now!

| iOS | Android |
|-|-|
|  |   |

**Comment by janicduplessis** [comment]
@motiz88 Nice just tested again and everything seems fine! Will try to have someone from fb have a look.

File: Libraries/Text/TextStylePropTypes.js [hunk 1]
```diff
@@ -49,7 +49,16 @@ const TextStylePropTypes = {
   textShadowRadius: ReactPropTypes.number,
   textShadowColor: ColorPropType,
   /**
-   * @platform ios
+   * Increase or decrease the spacing between characters. The default value is 0, for no extra
```
Comment (hramos) [inline Libraries/Text/TextStylePropTypes.js]: Can you add these docs to the docs at https://github.com/facebook/react-native-website instead? We'll be slowly pruning out these comment blocks over the next few months.

Comment (hramos) [reply Libraries/Text/TextStylePropTypes.js]: (It's fine to wait for this PR to be merged first)

**Comment by hramos** [comment]
Importing for internal review.

**Comment by hramos** [comment]
Sorry about the conflicting statements above - this still needs to be reviewed internally, but we'll need this to be rebased first in order to be able to import the PR cleanly.

**Comment by motiz88** [comment]
@hramos I've rebased this on top of `master` as requested.

**Comment by ejc3** [comment]
Agree this is something good to fix.

**Comment by zeh** [comment]
Any news on this? It seems a lot of effort was put into the PR, how else can we help this move forward?

**Comment by hramos** [comment]
The PR has already been imported, which is more than can be said for most PRs. It's now waiting for review from an employee. We'll sync back here if any changes are required.

**Comment by janicduplessis** [comment]
Thanks @motiz88 @hramos @mdvacca !! 🎉

**Comment by LouisJS** [comment]
Any news on this ?

**Comment by motiz88** [comment]
It's in 0.55:
https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md#055

**Pull request closed without merging**